[Merged by Bors] - Self-healing

lrettig opened the pull request:
## Motivation
See https://github.com/spacemeshos/SMIPS/issues/46

Requires #2394, #2393, #2357
Closes #2203
Closes #2687

## Changes
- adds zdist param: whereas hdist is the "Hare lookback" distance (the number of layers for which we consider hare results/local input vector rather than global opinion), zdist is the "Hare result wait" distance (the number of layers we're willing to wait for Hare to finish before reverting to invalidating layers with no input vector)
- Hare explicitly reports failed CPs to the mesh, with the results stored in memory

## Test Plan
TBD

## TODO
<!-- This section should be removed when all items are complete -->
- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [x] Finish rescoring block goodness after healing
- [x] Rescoring block unit test
- [x] Block voting weight unit test
- [x] Make sure healing -> verifying tortoise handoff is working
- [x] App test (for decay in active set size)
- [x] Split-then-rejoin test (two way split, three way split w/o majority fork)
- [x] Late blocks unit test(s)
- [x] Correctly weight votes from blocks with late atxs/those with a bad beacon value (will follow up in separate issue: #2540)
- [x] Multi tortoise unit tests

## DevOps Notes
<!-- Please uncheck these items as applicable to make DevOps aware of changes that may affect releases -->
- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [ ] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

lrettig committed:
Context and logging

Bring context down into verifying tortoise

lrettig committed:
Gofmt

lrettig committed:
More context and logging

lrettig committed:
Cleanup: logging and comments

lrettig committed:
Add support for explicitly invalid layers

Hare reports to mesh when it explicitly fails for a layer.
This allows tortoise to add votes against all blocks in such a layer.
Also adds some more minor cleanup.

lrettig committed:
Add zdist

Increase default hdist
Invalidate layers with no input vector older than zdist
Also clean up some interfaces

lrettig committed:
Gofmt

lrettig committed:
Merge branch 'develop' into healing

lrettig committed:
Minor cleanup

Comments

lrettig committed:
Some important logic fixes, and minor cleanup

- Set local opinion as opposed to all blocks in a layer if Hare failed
  for that layer.
- Correctly process exception votes when processing blocks (rather than
  summing base block vote + exception vote); handle abstain votes here
- Correctly interpret the lack of an explicit vote in support of a block
  in an older late as a vote against that block when summing global
  opinion

Lots of minor cleanup and improvements of comments, logs, variable names
Change some panics to errors

lrettig commented on tortoise/verifying_tortoise.go:
```diff
@@ -22,6 +23,14 @@ type blockDataProvider interface {
 	Retrieve(key []byte, v interface{}) (interface{}, error)
 }
 
+//func blockArrayToMap(ar []types.BlockID) (mp map[types.BlockID]struct{}) {
```
Remove

lrettig commented on tortoise/verifying_tortoise.go:
```diff
@@ -156,33 +173,35 @@ func (t *turtle) getSingleInputVectorFromDB(lyrid types.LayerID, blockid types.B
 		} else if exceptionBlock.LayerIndex < layerIndex {
-			t.logger.With().Debug("not marking block good because it points to block older than base block",
-				sourceBlockID,
+			// TODO: can a base block exception list point to blocks in the same layer, or must they all be newer?
```
Update with latest info from Tal

lrettig commented on tortoise/verifying_tortoise.go:
```diff
@@ -392,29 +441,35 @@ func (t *turtle) processBlock(block *types.Block) error {
-
 	for _, b := range block.ForDiff {
-		opinion[b] = opinion[b].Add(support.Multiply(t.BlockWeight(block.ID(), b)))
+		opinion[b] = support.Multiply(t.BlockWeight(block.ID(), b))
```
Check for collisions here? Maybe in malicious blocks

lrettig commented on tortoise/verifying_tortoise.go:
```diff
@@ -423,179 +478,180 @@ func (t *turtle) HandleIncomingLayer(newlyr *types.Layer) (pbaseOld, pbaseNew ty
-			t.logger.With().Warning("input vector abstains on all blocks", i)
+			if errors.Is(err, mesh.ErrInvalidLayer) {
+				// Hare already failed for this layer, so we want to vote against all blocks in the layer: an empty
+				// map will do the trick
```
List not map

lrettig commented on tortoise/verifying_tortoise.go:
```diff
@@ -423,179 +478,180 @@ func (t *turtle) HandleIncomingLayer(newlyr *types.Layer) (pbaseOld, pbaseNew ty
+		localLayerOpinionVec := t.inputVectorForLayer(layerBlockIds, rawLayerInputVector)
+		if len(localLayerOpinionVec) == 0 {
+			logger.With().Warning("no blocks in input vector, can't verify layer", candidateLayer)
 			break
```
Why break not continue?

lrettig committed:
Add exceptions back to hDist

Per https://github.com/spacemeshos/research/issues/33 we want to look
back to the start of the sliding window (currently defined as hdist)
when adding exceptions to a candidate base block

lrettig committed:
Add tortoise confidence param

Make comments clearer, some minor cleanup

lrettig committed:
Gofmt

lrettig committed:
Add initial criteria to trigger self-healing

lrettig committed:
Clarified point with Tal, remove TODO

lrettig committed:
Self-healing core logic, add weak coin stub

- Mostly finished self-healing core logic in tortoise
- Remove lots of unused code from block builder
- Adds weak coin stub to hare while waiting for #2393 to be merged

lrettig committed:
Cleanup weak coin stub

Change direction of data flow, hare -> mesh -> tortoise

lrettig committed:
Linting

lrettig committed:
Gofmt

lrettig committed:
Finish core healing logic

Improve some log output

lrettig committed:
Logic cleanup

Remove unused variables, clean up method signatures

lrettig committed:
Add explanatory comment

Regarding block and ATX weights, per Tal

lrettig committed:
Merge branch 'develop' into healing

lrettig committed:
Add WindowSize

lrettig committed:
Properly calculate opinion vector

Write a new function that properly calculates the most up to date
opinion of a layer, based on local or global opinion and the age of the
layer.

Lots of other cleanup

lrettig committed:
Minor logic cleanup

lrettig committed:
Logic cleanup

- Allow empty layers
- Never heal layers newer than hdist

lrettig committed:
Fix test

lrettig committed:
Gofmt

lrettig committed:
Periodically rerun tortoise from scratch

First draft

lrettig commented on tortoise/verifying_tortoise.go:
```diff
@@ -392,214 +450,406 @@ func (t *turtle) processBlock(block *types.Block) error {
+	logger := t.logger.WithContext(ctx).WithFields(layerID)
+
+	// for layers older than hdist, we vote according to global opinion
+	if layerID < t.Last-t.Hdist {
```
To do: what if this layer hasn't been verified yet?

lrettig replied:
Need to calculate global opinion

lrettig committed:
Significant cleanup to streamline tortoise API

- syncer no longer needs to maintain a validatingLayer, this hasn't been
  needed for a while (it's maintained in mesh)
- remove hare validation function (this does nothing but print a
  warning, which the mesh does too)
- change reporter to pass only a layerID, no blocks. downstream clients
  (including API handlers) can read layer data themselves as necessary.

lrettig committed:
Major tortoise logic cleanup

- Make tortoise logic more modular
- Separately handle incoming layers and incoming (late) blocks
- Clean up tortoise wrapper
- Prepare HandleLateBlocks to handle blocks in batches

lrettig committed:
Gofmt

lrettig committed:
First complete draft of tortoise state reversion

lrettig committed:
Linter

lrettig committed:
Merge branch 'develop' into healing

lrettig committed:
Merge branch 'develop' into healing

lrettig committed:
Concurrency fix for test

lrettig committed:
Explanatory comment

lrettig committed:
Another test fix

lrettig committed:
Don't apply state after hare

Only apply state after tortoise validates a layer

lrettig committed:
Bugfix, fix tests

lrettig committed:
Fix another test

to account for the fact that we no longer apply layer state after hare,
before tortoise.
Adds some more tortoise logging.

lrettig committed:
Bugfix, fix test

lrettig committed:
Don't run on privileged ports

lrettig committed:
Misc cleanup, tests, fixes

All existing tortoise unit tests pass
Still need to add new tests
Misc logging and code cleanup
Add some more logs
Some fixes on bounds checking, overflows

lrettig committed:
More cleanup of existing tests

Split out recovery test
Some cleanup based on tests
Move state into underlying tortoise struct for persist and recover

lrettig committed:
Fix another test concurrency issue

lrettig commented:
bors try

lrettig committed:
WIP new tortoise tests

lrettig commented on tortoise/verifying_tortoise.go:
```diff
@@ -261,68 +314,95 @@ func (t *turtle) BaseBlock() (types.BlockID, [][]types.BlockID, error) {
-				t.logger.Panic("can't find old layer %v", i)
+				// this should not happen! we only look at layers up to the last processed layer, and we only process
+				// layers with valid block data.
+				logger.Error("no block ids for layer in database")
```
print err

lrettig commented on tortoise/verifying_tortoise_test.go:
```diff
@@ -361,65 +368,221 @@ func TestTurtle_Recovery(t *testing.T) {
+	}()
+
+	getHareResults := func(l types.LayerID) ([]types.BlockID, error) {
+		return mdb.LayerBlockIds(l)
```
Is this necessary? Do these methods have different signatures? Looks like no. Remove the wrapper

lrettig committed:
Complete layerOpinionVector test

Some misc cleanup, logic tweaks

lrettig committed:
Fleshing out tests

lrettig committed:
Add two more tests

lrettig committed:
Add another tortoise test

Add a TODO about a case we still need to handle: voting against blocks a
base block supports that we disagree with

lrettig committed:
Finish test

lrettig committed:
Add another test

lrettig committed:
Add two more tests

lrettig committed:
Verify layers test WIP

lrettig committed:
Finish verify layers test

Includes first test of self-healing code path
Some small logic fixes

lrettig committed:
Clean up some layer arithmetic

lrettig commented on tortoise/verifying_tortoise.go:
```diff
@@ -362,244 +450,497 @@ func (t *turtle) persist() error {
+
+		// it's possible that self-healing already validated a layer
+		if t.Verified >= candidateLayerID {
+			logger.Info("self-healing already validated this layer")
```
verify not validate

lrettig commented on tortoise/verifying_tortoise.go:
```diff
@@ -362,244 +450,497 @@ func (t *turtle) persist() error {
-				break layerLoop
+				// short-circuit processing of this layer, but allow verification of later layers to continue after
+				// self-healing has finished
+				break candidateLayerLoop
```
continue not break

lrettig committed:
Add back TODO

lrettig committed:
Merge branch 'develop' into healing

lrettig committed:
Improve test verify layers

Make test more thorough and robust

lrettig committed:
Finish process new blocks test

lrettig committed:
WIP self-healing tests

Refactor tortoise code to decouple handling incoming blocks/layer data
from attempting to verify layers, for purposes of testing
Clean up test temp dir and cleanup

lrettig committed:
Refactor last layer for verification

No need to pass this around, just use state instead. Makes the logic
simpler.

Fix the math for calculating delta for layer verification

lrettig committed:
Always count votes up to t.Last

lrettig committed:
Comments

lrettig committed:
Update test to match new params

lrettig committed:
Healing tests WIP

lrettig committed:
Add global and local threshold params

Rewrite calculateGlobalOpinion to use them

lrettig committed:
Finish logic updates per Tal's instructions

lrettig committed:
Linter

lrettig committed:
Fix previously broken old tests

lrettig committed:
Fix eviction

lrettig committed:
Fix test layer opinion test

lrettig committed:
Fix calculate exceptions test

One small logic tweak

lrettig committed:
WIP updating TestVerifyLayers

rewriting the self-healing test cases to match the new logic

lrettig committed:
Merge branch 'develop' into healing

lrettig committed:
Merge branch 'develop' into healing

Properly print hare tick errors

lrettig committed:
Resolve merge conflict in new test

lrettig committed:
Small logic fix, finish another test

Tortoise won't attempt to continue verification after healing has failed
Finish test that tests handoff from ordinary tortoise -> healing ->
ordinary tortoise

lrettig committed:
Finish a couple more tests

Add more healing tests

lrettig committed:
Cleanup

lrettig committed:
Merge branch 'develop' into healing

lrettig committed:
Merge branch 'develop' into healing

lrettig committed:
Fix test

lrettig committed:
Merge branch 'develop' into healing

lrettig committed:
Fix eviction tests

lrettig committed:
Fix remaining tortoise tests

Improve errmsg

lrettig committed:
Minor cleanup

lrettig committed:
Improve and clean up tests for readability

lrettig committed:
Add new layer pattern test

lrettig committed:
Add new healing pattern test

lrettig committed:
Bugfix: init map in mem mesh DB

lrettig committed:
Better logging, new test WIP

lrettig committed:
Merge branch 'develop' into healing

lrettig committed:
Use local opinion when marking blocks good

rather than always relying on just input vector. For blocks in older
layers, and layers that have been healed, this will allow blocks to be
marked good again. Per Tal.

lrettig committed:
Add test for exiting healing regime

Currently failing until we figure out what to do about blocks previously
marked not good

lrettig committed:
Add test for reversion

Fix a couple of small bugs

lrettig committed:
Move TODO to new syncer code

lrettig committed:
Merge branch 'develop' into healing

lrettig committed:
WIP adding new partition test

lrettig committed:
Finish new test

lrettig committed:
Misc, minor cleanup of existing tests

lrettig committed:
WIP adding healing app test

lrettig committed:
Improve app test

Check verified layer not processed layer

lrettig committed:
Improve logging, clock, turbohare

Fix a concurrency bug in turbohare
Reduce log verbosity for syncer and elsewhere
Introduce a shared clock

lrettig committed:
Increase timeout, reduce log verbosity

lrettig committed:
Rename inputvector to input vector for consistency

lrettig committed:
Finish updating app test, logging improvements

Finish adding basic support for healing to app test. It now supports
calculating ATXs, weight, block count, etc. after some nodes are killed.
Some misc logging improvements

lrettig committed:
Don't write data on already-evicted layers

Rename a few things for clarity
Expand eviction test

lrettig committed:
Clean up weight calculation in test

Ensure that the verified layer advances at least two layers before
passing test

lrettig committed:
Bugfix in weight calculation

lrettig committed:
WIP new balance attack test

lrettig committed:
Finish balance attack test

Update some comments and add some TODOs accordingly

lrettig committed:
WIP: rescore block goodness after healing

lrettig committed:
Merge branch 'develop' into healing

lrettig committed:
Improve how we call block scoring method

Rename it, and only call it in one place post-healing

lrettig committed:
Issue link

lrettig committed:
Weight block votes using ATX weight

lrettig committed:
Merge branch 'develop' into healing

lrettig committed:
Add TestDetermineBlockGoodness unit test

lrettig committed:
Add new TestScoreBlocks test

lrettig committed:
Replace persistent mesh with in mem mesh

For most tests

lrettig committed:
Add TestVoteWeight test

lrettig committed:
Rename/restructure some tests

lrettig committed:
Add weighted ATX to tests, TestLateBlocks test

lrettig committed:
Merge branch 'develop' into healing

lrettig committed:
Misc cleanup

Remove some unused code

lrettig committed:
Split atx interfaces for prod and tests

lrettig committed:
Minor cleanup

Related to tortoise beacon

lrettig committed:
Tortoise beacon minor cleanup

lrettig committed:
Add layer clock for ATX timeliness

lrettig committed:
Minor cleanup

Remove unused ProcessAtxs method

lrettig committed:
Minor cleanup

lrettig committed:
Merge branch 'develop' into healing

lrettig committed:
Continue resolving merge conflict

lrettig committed:
Continue to resolve merge conflict

Make linter happy

lrettig committed:
Variable naming

lrettig committed:
Bugfixes in tortoise tests

Prevent overflow in lookback
Properly set atxdb in trtl

lrettig committed:
Bugfix: vector arithmetic

Properly detect and handle overflow/wraparound

lrettig committed:
Bugfixes, test fixes

lrettig committed:
Merge branch 'develop' into healing

lrettig committed:
Fix TestGetLocalBlockOpinion

Update to match current logic: return a binary support or against (no
abstain)

lrettig committed:
Bugfix in vector math, clean up tests

Properly set block weights using ATX in all cases

lrettig committed:
Minor logic tweaks

More cleanup from merge

lrettig committed:
Clean up vector math, add tests

lrettig committed:
Add TestVoteVectorForLayer

lrettig committed:
Continue working on layer pattern tests

Improve debugging and add some TODO items
Set atx timestamps more carefully in tests
Allow empty input vector fn when making layers in tests

noamnelke commented on tortoise/util.go:
```diff
@@ -21,15 +20,36 @@ func (a vec) Field() log.Field {
+	if x != 0 && one/x != a[0] || two/x != a[1] {
+		panic("vector arithmetic overflow")
+	}
 	return a
```
You're never actually doing the multiplication, only checking for overflow and then discarding the result.

You're missing either (in-place multiplication):
```go
a[0] = one
a[1] = two
```
or returning a copy of the vector with the multiplication result.

noamnelke commented on tortoise/util.go:
```diff
@@ -21,15 +20,36 @@ func (a vec) Field() log.Field {
+	diff := one - two
+	if diff > one {
+		panic("vector arithmetic overflow")
+	}
```
A `uint64` isn't guaranteed to fit into an `int64` which is why you need a check here, but the check you do isn't enough. If `two` overflows it will be negative, causing `diff` to be greater than `one`, but if `one` overflows and `two` doesn't the method won't panic.

You can separately check if each component overflows before doing the subtraction, but that would mean that we just don't support any values in vectors that don't fit in a `uint32` - so a better solution would be to change to `uint32` in vectors and fail when first trying to place an overflowing value there. Then the conversion from `uint32` to `int64` is guaranteed to succeed and no check is necessary.

Alternatively, we can do a calculation that only overflow when the result doesn't fit in an `int64` (solution 1) or return a `uint64` + sign bit (solution 2).

Solution 1 (no code changes outside `netVote`):
```go
abs := uint64()
sign := 1
if a[0] > a[1] {
	res = a[0] - a[1]
} else {
	res = a[1] - a[0]
	sign = -1
}
if int64(res) < 0 {
	panic("vector arithmetic overflow")
}
return int64(res) * sign
```

Solution 2 (`netVote` returns `absNetVote` + `positiveNetVote` - requires code changes, but cannot overflow):
```go
if a[0] > a[1] {
	return a[0] - a[1], true
} else {
	return a[1] - a[0], false
}
```

lrettig replied:
This code has been updated. Let's continue the conversation in #2629.

noamnelke commented on tortoise/util.go:
```diff
@@ -53,7 +73,7 @@ func (a vec) String() string {
 func calculateOpinionWithThreshold(logger log.Log, v vec, layerSize int, theta uint8, delta float64) vec {
 	threshold := float64(theta) / 100 * delta * float64(layerSize)
-	netVote := float64(v[0] - v[1])
+	netVote := float64(v.netVote())
```
How is it ok that we use floats here? I thought that they are unacceptable in consensus code due to varying implementation / behaviour across processor architectures 🤔

lrettig replied:
Let's continue the conversation in #2629

lrettig committed:
Continue to improve layer patterns tests

Make logic more dynamic
Combine two tests
Test is passing, indicating that, after the expected number of layers,
healing does hand off back to verifying tortoise

lrettig committed:
Merge branch 'develop' into healing

lrettig committed:
Fix two more tests

Re-add some test code that was lost in merge conflict

lrettig committed:
Cleanup app startServices

lrettig committed:
Merge branch 'develop' into healing

lrettig committed:
Streamline a few error messages

lrettig committed:
Streamline a couple more log messages

lrettig committed:
Properly shut down peer manager

lrettig committed:
Fix race condition in fetcher shutdown

lrettig committed:
Streamline more error messages

lrettig committed:
Make linter happy

lrettig committed:
Fix race condition in p2p sim

Use nonblocking send

lrettig committed:
Streamline log messages

lrettig committed:
Fix panic on zero peer count

Print errors instead, when this occurs due to a race condition
Streamline some more logging

lrettig committed:
Merge branch 'develop' into healing

lrettig committed:
Make linter happy

lrettig committed:
Streamline some more log messages

lrettig committed:
Properly close NodeService PeerCounter

lrettig committed:
Fix race conditions in discovery

Use nonblocking send in ping, getaddress

lrettig committed:
Fix race condition in p2p sim

Perform nonblocking send to send direct message, in case listener has
disappeared

lrettig committed:
Merge branch 'develop' into healing

lrettig committed:
Print whole stack in TestShutdown

lrettig committed:
Merge branch 'develop' into healing

lrettig committed:
Continue resolving merge conflicts

Make linter happy

lrettig committed:
Merge branch 'develop' into healing

lrettig committed:
Fix TestPropagationEventLoop

lrettig committed:
Fix race condition in peers listener

And properly print stack in TestShutdown

lrettig committed:
Improve app test logging

lrettig committed:
Reduce log level of one message

lrettig committed:
Rename some fields for clarity

lrettig committed:
Make hare params clearer and shorter

lrettig committed:
Misc cleanup

Improve some comments
Rename some stuff for clarity
Add TODOs
Improve log messages

lrettig committed:
Improve trtl logging and comments

lrettig committed:
Bugfix in calculateExceptions

Need to look at simplified vote
Adds warning

lrettig committed:
Pass simulated hare output to mesh

lrettig commented:
bors try

lrettig committed:
Merge branch 'develop' into healing

lrettig committed:
Add regression test for vote weight bugfix

lrettig committed:
Misc minor cleanup

lrettig committed:
Bring back different node weights in app test

lrettig committed:
Make sure healing kicked in in app test

lrettig committed:
Whitespace

lrettig committed:
Misc cleanup

Minor logic tweaks, improved logging

lrettig committed:
Bugfix: don't double-count block vote weights

lrettig committed:
Fix healing app test

lrettig committed:
Tweak errmsg

lrettig committed:
Merge branch 'develop' into healing

lrettig committed:
Make linter happy

lrettig committed:
Handle errors in tests

lrettig committed:
Tweak tests

lrettig committed:
Merge branch 'develop' into healing

lrettig committed:
Add regression test for block weight bugfix

lrettig committed:
Variable naming

lrettig commented:
bors try

lrettig committed:
Add test for HandleLateBlocks

lrettig committed:
TestMultiTortoise WIP

lrettig committed:
TestMultiTortoise first working version

lrettig committed:
Add weight arg to generateBlocks

lrettig committed:
Partition test WIP

lrettig commented:
bors try

lrettig committed:
First partition test working

lrettig committed:
Add 50/50 partition test

lrettig commented:
bors try

lrettig committed:
Fix go version in release workflow

lrettig committed:
Add test for three-way split

lrettig committed:
Do not explicitly vote against bad blocks

Instead, implicitly vote against them by not voting for them. Per Tal.
We still need to add support for explicitly voting against bad blocks
that the base block supports.

lrettig committed:
Rejoin test WIP

y0sher commented on hare/hare.go:
```diff
@@ -61,7 +62,7 @@ type Hare struct {
 	sign Signer
 
-	msh     meshProvider
+	mesh    meshProvider
```
`msh` was used because `mesh` is a package name hence can be confusing.

lrettig replied:
I see your point, but I'd rather optimize for readability. That package is not imported here, so there is no conflict, at least in this module. We could probably clean this up globally.

y0sher commented on hare/hare.go:
```diff
@@ -156,20 +165,14 @@ var ErrTooLate = errors.New("consensus process finished too late")
-	// check validity of the collected output
-	if !h.validate(blocks) {
-		h.WithContext(ctx).Error("failed to validate the collected output set")
+		blocks = append(blocks, v)
```
at least in older versions of go, creating the slice and assigning each index when the size is known was more efficient than append. though I get append is more readable.

y0sher commented on tortoise/util.go:
```diff
@@ -1,79 +1,95 @@
+	a[0] += v[0]
+	a[1] += v[1]
+	// prevent overflow/wraparound
+	if a[0] < v[0] || a[1] < v[1] {
```
maybe we could check before adding instead of check that overflow has happened?

lrettig replied:
not sure what you mean by "check before adding", what's the difference? you mean like turn this into an error in the code that calls Add?

y0sher replied:
I was thinking preventing the overflow before it happened but I know see there'
s not much difference. resolving this.

y0sher commented on tortoise/verifying_tortoise.go:
```diff
@@ -36,11 +61,38 @@ func blockMapToArray(m map[types.BlockID]struct{}) []types.BlockID {
+	bdp   blockDataProvider
+	clock layerClock
+
+	// TODO: can some or all of these be unexported? is this an issue for serialization?
```
it is a serialization issue, w/e you want to be serialized (at least with current xdr impl) has to be exported.

lrettig replied:
got it, I thought so, thanks. updated the comment accordingly.

y0sher commented on tortoise/verifying_tortoise.go:
```diff
@@ -49,73 +101,132 @@ type turtle struct {
 
-func (t *turtle) init(genesisLayer *types.Layer) {
+// cloneTurtle creates a new verifying tortoise instance using the params of this instance
+func (t *turtle) cloneTurtle() *turtle {
```
the name got me confused at first, thought you were cloning the tables as well, maybe mention you're cloning or copying configuration in the name.

y0sher commented on tortoise/verifying_tortoise.go:
```diff
@@ -130,19 +241,23 @@ func blockIDsToString(input []types.BlockID) string {
+// returns the binary local opinion on the validity of a block in a layer (support or against).
+// abstain is not an option here: if the current opinion on a layer is abstain, an opinion on
+// a block in that layer will be against.
+// TODO: does this ever need to return abstain?
```
if we know hare wasn't terminated yet we should return abstain, this is why the error case is here. I assumed a terminated hare with empty set is perfectly valid so no err but if there are no results return an error to indicate it's still running.

lrettig replied:
fixed this, thanks for highlighting

y0sher commented on tortoise/verifying_tortoise.go:
```diff
@@ -194,428 +309,936 @@ func (t *turtle) checkBlockAndGetInputVector(
+				logger.With().Debug("added vote diff",
+					log.FieldNamed("diff_block", bid),
+					log.String("diff_class", voteClass))
+				if layerID.Before(baseBlockLayerID) {
```
should we allow this to happen? why not add the diff only if the layer is newer? it sounds to me it might be required for self healing.

lrettig replied:
for a given base block candidate, it's essential that the diffs be added, that's how the verifying tortoise algorithm works (otherwise, the votes would be counted incorrectly). what we _could_ do instead is do a better job of picking base blocks, so as to minimize the length of the exception list. that's a to do item in #2402.

y0sher commented on tortoise/verifying_tortoise.go:
```diff
@@ -194,428 +309,936 @@ func (t *turtle) checkBlockAndGetInputVector(
+		// We do not explicitly add votes against blocks that the base block does _not_ support, since by not voting to
+		// support a block in an old layer, we are implicitly voting against it. But if the base block does explicitly
+		// support a block and we disagree, we need to add a vote against here.
+		// TODO: this is not currently possible since base block opinions aren't indexed by layer. See
```
you can always get full block from db and check the layer though it's quite a lot of overhead.

lrettig replied:
yeah, I'm afraid that would be too expensive. this is a high priority followup item and will be addressed in #2424.

y0sher commented on tortoise/verifying_tortoise.go:
```diff
@@ -194,428 +309,936 @@ func (t *turtle) checkBlockAndGetInputVector(
+// whether or not the block's ATX was received on time, and on how old the layer is.
+// TODO: for now it's probably sufficient to adjust weight based on whether the ATX was received on time, or late, for
+//   the current epoch. See https://github.com/spacemeshos/go-spacemesh/issues/2540.
+// TODO: does blockVotedOn ever matter here? can we remove it?
```
unless you see some future use, remove blockVotedOn, I assumed it's needed when creating the stub but it was just a guess.

lrettig replied:
removed

y0sher commented on tortoise/verifying_tortoise.go:
```diff
@@ -194,428 +309,936 @@ func (t *turtle) checkBlockAndGetInputVector(
 
-func (t *turtle) processBlock(block *types.Block) error {
+func (t *turtle) processBlock(ctx context.Context, block *types.Block) error {
+	logger := t.logger.WithContext(ctx).WithFields(
```
won't this add double key fields when logging line L563 or L564?

lrettig replied:
yup, but I think it's a small price to pay for coherence in the logs throughout the code block

y0sher commented on tortoise/verifying_tortoise.go:
```diff
@@ -194,428 +309,936 @@ func (t *turtle) checkBlockAndGetInputVector(
+		return baseVote.Multiply(weight), nil
 	}
 
+	// TODO: this logic would be simpler if For and Against were a single list
```
agreed

lrettig replied:
also a high priority follow up item, in #2369

lrettig committed:
Merge branch 'develop' into healing

lrettig committed:
Finish rejoin test

y0sher commented on tortoise/verifying_tortoise.go:
```diff
@@ -194,428 +309,936 @@ func (t *turtle) checkBlockAndGetInputVector(
+		}
+
+		// add base block vote only if there weren't already exceptions support/against/abstain for this block
+		if _, exists := opinion[blk]; !exists {
```
I was under the impression that the baseblock votes are "added" implicitly.. 
also, do we need to weight these votes as well?

lrettig replied:
this is precisely the operation where the base block votes are implicitly added to the block's vote! the base block's opinions should already have had the weight factored in here.

actually, this raises an interesting question: if the base block has weight X and a block based on it has weight Y, what weight should we use for the "implicit" opinion of X that's included in Y's vote? I'll ask Tal.

lrettig replied:
#2687

y0sher commented on tortoise/verifying_tortoise.go:
```diff
@@ -194,428 +309,936 @@ func (t *turtle) checkBlockAndGetInputVector(
-			// (2) all diffs appear after the base block and are consistent with the input vote vector
-			t.logger.With().Debug("marking good block", b.ID(), b.LayerIndex)
+		if err := t.processBlock(ctx, b); err != nil {
+			logger.With().Error("error processing block", log.Err(err))
```
should blocks that we had errors processing be added to `filteredBlocks`? if not we should probably continue here.

y0sher commented on tortoise/verifying_tortoise.go:
```diff
@@ -194,428 +309,936 @@ func (t *turtle) checkBlockAndGetInputVector(
+	numGood := 0
+	for _, b := range blocks {
+		if t.determineBlockGoodness(ctx, b) {
+			// note: we have no way of warning if a block was previously marked as not good
```
indeed we can't check the list, but if we received this block previously and it isn't in the list, doesn't it mean we thought it wasn't good?

lrettig replied:
if we received and processed the block previously, yes. but I don't think we have an easy way of checking that here, either...

y0sher commented on tortoise/verifying_tortoise.go:
```diff
@@ -194,428 +309,936 @@ func (t *turtle) checkBlockAndGetInputVector(
+		logger.Debug("base block is not good")
+	} else if baseBlock, err := t.bdp.GetBlock(block.BaseBlock); err != nil {
+		logger.With().Error("inconsistent state: base block not found", log.Err(err))
+	} else if true &&
```
I don't see why you need this `true` here?
also I know it's a style preference but I think 2 ifs and returns are clearer than `if else if else if`

lrettig replied:
it's also a style thing, to keep the three conditionals on the following lines aligned

y0sher commented on tortoise/verifying_tortoise.go:
```diff
@@ -194,428 +309,936 @@ func (t *turtle) checkBlockAndGetInputVector(
+						// TODO LANE: this operation can be very expensive and very slow after a few layers,
+						//   since no caching is currently performed. it needs to be optimized and cached.
+						if err := t.scoreBlocksByLayerID(ctx, layerID); err != nil {
+							// if we fail to process a layer, there's probably no point in trying to recore blocks
```
```suggestion
							// if we fail to process a layer, there's probably no point in trying to rescore blocks
```

lrettig committed:
Merge branch 'develop' into healing

lrettig committed:
Revert accidental GoLand config change

Added inadvertantly as part of a merge commit

lrettig committed:
Handle an unhandled error in meshdb

lrettig committed:
Comments and todo issue links

lrettig committed:
Surface abstain from getLocalBlockOpinion

Updating tests WIP

lrettig committed:
Minor cleanup, add TODO links

lrettig committed:
Finish allowing abstain local layer opinion

Update tests accordingly: it now takes a bit longer to heal since we
start with an initial vote of 0 (abstain), so it takes longer to catch
up and exceed the threshold

lrettig committed:
Make linter happy

lrettig commented:
bors try

lrettig commented:
bors try

lrettig committed:
Parameterize tortoise rerun interval

y0sher commented on tortoise/verifying_tortoise.go:
```diff
@@ -130,492 +242,1011 @@ func blockIDsToString(input []types.BlockID) string {
+					//   anything we can do about this? e.g., explicitly pick base blocks that agree with the new
+					//   opinion, or pick older base blocks.
+					//   see https://github.com/spacemeshos/go-spacemesh/issues/2678
+					if coin, exists := t.bdp.GetCoinflip(ctx, t.Last.Sub(1)); exists {
```
shouldn't we get the coinflip for the current `layerID` we're checking ?

lrettig replied:
nope, we use the coinflip for the top layer we've seen. I clarified this with Tal. I'll add an explanatory comment to this effect.

y0sher commented on tortoise/verifying_algorithm.go:
```diff
@@ -11,52 +13,104 @@ import (
+	WindowSize      uint32 // tortoise sliding window: how many layers we store data for
+	Log             log.Log
+	Recovered       bool
+	RerunInterval   time.Duration // how often to rerun from genesis
```
`RerunInterval` is only used in `tortoise/verifying_algorithm.go` but it's still part of `turtle`. is it because we want to persist this setting?

lrettig replied:
yup

y0sher commented on mesh/meshdb.go:
```diff
@@ -324,17 +331,36 @@ func (m *DB) ContextualValidity(id types.BlockID) (bool, error) {
+// method lets us know that Hare has given up.
+func (m *DB) InvalidateLayer(ctx context.Context, layerID types.LayerID) {
+	m.WithContext(ctx).With().Info("recording hare invalidated layer in mesh", layerID)
+	m.invalidatedLayers[layerID] = struct{}{}
```
this map is accessed from multiple code paths, shouldn't it be thread-safe?

lrettig replied:
it's only accessed in one place, the Hare's `outputCollectionLoop`, and there should only be one of those running at a time. do you think we still need to protect it?

y0sher reviewed (approved):
Good job, I especially liked how detailed the documentation is, it made reviewing this a breeze. I couldn't find anything major so I'm approving. please look at my comments and consider them, they are mainly suggestions/questions.

lrettig commented:
bors try

lrettig commented:
bors try

lrettig commented:
bors try

lrettig commented:
bors try

lrettig commented:
bors try

lrettig committed:
Factor in review feedback

Some minor cleanup and improvements
Add some explanatory comments and more TODO issue links
Rename trtl.cloneTurtle for clarity

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig committed:
Re-weight implicit base block votes

Using voting block's weight rather than base block weight

lrettig committed:
Merge branch 'develop' into healing

lrettig commented:
bors merge

countvonzero commented on syncer/syncer.go:
```diff
@@ -268,12 +269,12 @@ func (s *Syncer) synchronize(ctx context.Context) bool {
 		<-vDone
 		s.setStateAfterSync(ctx, success)
-		logger.With().Info(fmt.Sprintf("finished sync run #%v", s.run),
+		logger.With().Debug(fmt.Sprintf("finished sync run #%v", s.run),
```
i'd rather this to be the Info and line 304 to be Debug.

countvonzero replied:
but no fuss. i can change it after you merge

lrettig replied:
done!

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig committed:
Update syncer log levels

lrettig commented:
bors merge

dshulyak commented:
please use logtest.New(t) in tests and not log.NewDefault

lrettig committed:
Merge branch 'develop' into healing

lrettig committed:
Use logtest.New instead of log.NewDefault in tests

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig committed:
Fix initial map size

lrettig committed:
Add deep eviction test

lrettig committed:
Update eviction test math

We allow opinions on blocks up to WindowSize+3 prior to parent block
layer

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig commented:
bors merge

On Sat, Aug 21, 2021, 12:20 bors[bot] ***@***.***> wrote:

> Build failed:
>
>    - ci-stage2
>    <https://github.com/spacemeshos/go-spacemesh/runs/3389889106>
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spacemeshos/go-spacemesh/pull/2400#issuecomment-903139973>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZJWNGETJMQ2ISNAN4O7KTT57G5XANCNFSM43XVGWUQ>
> .
>

lrettig commented:
bors merge

On Sat, Aug 21, 2021, 13:26 bors[bot] ***@***.***> wrote:

> Build failed:
>
>    - ci-stage2
>    <https://github.com/spacemeshos/go-spacemesh/runs/3390072515>
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spacemeshos/go-spacemesh/pull/2400#issuecomment-903148718>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZJWNDZECJXIQ2IZ7PFMALT57OVPANCNFSM43XVGWUQ>
> .
>

lrettig commented:
bors merge

On Sat, Aug 21, 2021, 14:39 bors[bot] ***@***.***> wrote:

> Build failed:
>
>    - ci-stage2
>    <https://github.com/spacemeshos/go-spacemesh/runs/3390294278>
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spacemeshos/go-spacemesh/pull/2400#issuecomment-903158716>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZJWNENNPP4GEQ4EYUX7STT57XENANCNFSM43XVGWUQ>
> .
>

lrettig committed:
Merge branch 'develop' into healing

lrettig commented:
bors merge

lrettig commented:
bors merge

On Sat, 21 Aug 2021 at 20:23, bors[bot] ***@***.***> wrote:

> Build failed:
>
>    - ci-stage2
>    <https://github.com/spacemeshos/go-spacemesh/runs/3391292157>
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spacemeshos/go-spacemesh/pull/2400#issuecomment-903192919>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZJWNHHUZMWTXAJF2FTKFDT6A7PNANCNFSM43XVGWUQ>
> .
>

lrettig commented:
bors merge

lrettig committed:
Merge branch 'develop' into healing

lrettig commented:
bors merge

lrettig committed:
Minor cleanup logging and comments

Remove redundant code

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig committed:
Minor log improvements

lrettig commented:
bors merge

dshulyak commented on tortoise/verifying_tortoise.go:
```diff
@@ -130,492 +242,1007 @@ func blockIDsToString(input []types.BlockID) string {
+
+	// check if the ATX was received on time
+	// TODO: add an exception for sync, when we expect everything to be received late
+	if atxTimestamp.Before(nextEpochStart) {
```
@lrettig this appears to be the reason why the verified layer was stuck on my new devnet. it might be as well the reason for latenodes failure, or i misunderstood something?

lrettig replied:
good point, I'm testing without this

lrettig committed:
Merge branch 'develop' into healing

lrettig committed:
Temporarily disable ATX timeliness check

This may be causing sync failures

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig committed:
Minor logging/context improvement

lrettig committed:
Don't reapply oldPbase state

lrettig committed:
Logs and comments

lrettig committed:
Correctly configure block listener logger

lrettig commented:
Here are the logs for the latest `latenodes` systest failure (with some filters): http://34.83.138.105:5601/goto/760e4d8aa78c69d21358608525d7ffb9

Here are the first few relevant errors/warnings (with filters applied): http://34.83.138.105:5601/goto/91a1cc5a33cfa8ec699fa98c69f381d3

I'm seeing the following errors and warnings:
- `hare | error checking if identity is active | empty active set for layer 6 in epoch 2`
- `blockBuilder | failed to check for block eligibility | failed to get latest atx for node in epoch 2: failed to get atx id for target epoch 2: atx for node 08d24 with publication epoch 1: leveldb: not found`
- `hareOracle | error getting active set | empty active set for layer 6 in epoch 2`
- `blockOracle | failed to calculate eligibility proofs | failed to get latest atx for node in epoch 2: failed to get atx id for target epoch 2: atx for node 08d24 with publication epoch 1: leveldb: not found`
- `mesh | requested unknown layer hash`
- `tBeacon | failed to send proposal message | proposalPassesEligibilityThreshold: atxThreshold: zero epoch weight provided`
- `tBeacon | failed to handle proposal message | get node ATXID for epoch (miner ID 78bc0): atx for node 78bc0 with publication epoch 1: leveldb: not found`
- `tBeacon | failed to handle first voting message | verify first voting message: get node ATXID for epoch (miner ID f57a2): atx for node f57a2 with publication epoch 1: leveldb: not found`
- `hare | error processing hare tick | not starting hare since node is not synced`

It appears we still have issues syncing ATXs.

@countvonzero @nkryuchkov I could use your help troubleshooting the sync and beacon related issues here.

lrettig committed:
Bugfix: don't fail hare when not synced

This had the unfortunate consequence of causing tortoise to subsequently
vote _against_ all blocks in layers that arrived while syncing. Oof.

Spotter: dshulyak

lrettig committed:
Minor logging/logic cleanup

lrettig commented:
bors merge

lrettig committed:
Fix test

lrettig commented:
bors merge

lrettig committed:
Fix test

lrettig commented:
bors merge

lrettig committed:
Bugfix in mesh layer arithmetic

Should've been included in 1fa05a3a

lrettig commented:
bors merge

dshulyak commented:
bors merge

dshulyak commented:
weird. different unit test is failing every time, it didn't even come to latenodes in last 4 tries

bors merge

dshulyak commented:
bors merge

dshulyak commented:
i don't understand why unit tests are so flaky, there are now 3 failures in a single run:
- TestMesh_AddBlockWithTxs_PushTransactions_UpdateUnappliedTxs
- TestMesh_updateStateWithLayer
- TestFindNodeProtocol_FindNode_Concurrency
i just restarted them on develop in my branch and they passed from the first try https://github.com/spacemeshos/go-spacemesh/pull/2683

bors merge

dshulyak commented:
bors merge

dshulyak commented:
There must be something in this branch that makes those tests crash consistently on CI

lrettig committed:
Fix mesh test

Layer arithmetic

lrettig committed:
Fix test

More layer arithmetic fixes

lrettig committed:
Merge branch 'develop' into healing

lrettig commented:
bors merge

lrettig commented:
> i don't understand why unit tests are so flaky, there are now 3 failures in a single run:
> 
> * TestMesh_AddBlockWithTxs_PushTransactions_UpdateUnappliedTxs
> * TestMesh_updateStateWithLayer
> * TestFindNodeProtocol_FindNode_Concurrency
>   i just restarted them on develop in my branch and they passed from the first try [Refactor trtl.Persist to use its own store and do the task more efficiently #2683](https://github.com/spacemeshos/go-spacemesh/pull/2683)

These failures were due to a bugfix I made in the mesh code, just needed to update those tests - `TestFindNodeProtocol_FindNode_Concurrency` is flaky but the others were not a flakiness issue

countvonzero commented on mesh/reward_test.go:
```diff
@@ -211,57 +211,85 @@ func TestMesh_updateStateWithLayer(t *testing.T) {
+
 	// test states are the same when one input is from tortoise and the other from hare
-	assert.Equal(t, s.Txs, s2.Txs)
+	require.NotEqual(t, s.Txs, s2.Txs)
```
@lrettig 
is this correct? if we want the same result, it should be equal right?

lrettig replied:
the comment is a little misleading (see identical comment a few lines down), but the change is intentional and this should be correct. all layers _except_ the latest layer have been applied here, so the state should not be the same. then the latest layer is applied, and then the state should be the same. that's what this is testing.

countvonzero commented on mesh/reward_test.go:
```diff
@@ -211,57 +211,85 @@ func TestMesh_updateStateWithLayer(t *testing.T) {
+	mesh2.ValidateLayer(context.TODO(), l)
+
+	// test states are the same when one input is from tortoise and the other from hare
+	require.Equal(t, s.Txs, s2.Txs)
```
why not do `for i := 0; i < numOfLayers; i++` in line 228 and `require.Equal(t, s.Txs, s2.Txs)` directly?
why care about this one last layer difference?

lrettig replied:
the test is more thorough if it isolates the final layer application: the states are _not_ equal before the final layer is applied, and they _are_ equal after.
